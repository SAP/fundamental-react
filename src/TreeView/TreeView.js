import PropTypes from 'prop-types';
import Tree from './_Tree';
import TreeBranch from './_TreeBranch';
import TreeCol from './_TreeCol';
import TreeHead from './_TreeHead';
import TreeItem from './_TreeItem';
import TreeRow from './_TreeRow';
import React, { Component } from 'react';

class TreeView extends Component {
    constructor(props) {
        super(props);

        this.state = {
            expandData: {},
            isExpandAll: false
        };
    }

    static getDerivedStateFromProps(props, state) {
        const {
            isExpandAll,
            expandData: expandDataProp
        } = props;

        if (typeof expandDataProp !== 'undefined') {
            // If user provides a prop value for expandData then set that in state
            return {
                ...state,
                expandData: expandDataProp
            };
        }

        if (typeof isExpandAll !== 'undefined') {
            // If user provides a prop value for isExpandAll then update state accordingly
            const {
                expandData
            } = state;
            const newExpandData = {};

            // Expand/Collapse all rows
            Object.keys(expandData).forEach((rowId) => newExpandData[rowId] = isExpandAll);

            return {
                ...state,
                expandData: newExpandData,
                isExpandAll
            };
        }

        return state;
    }

    // Callback for each TreeRow to toggle expand/collapse state
    toggleExpand = (rowId) => {
        this.setState((prevState) => {
            const {
                expandData
            } = prevState;
            const {
                onExpandChange
            } = this.props;

            // If value doesn't exist, initialize it to false
            const newValue = (rowId in expandData) ? !expandData[rowId] : false;

            const newExpandData = {
                ...expandData,
                [rowId]: newValue
            };

            // Callback function
            onExpandChange(newExpandData);

            return {
                expandData: newExpandData
            };
        });
    }

    // Callback for TreeHead to toggle expand/collapse all state
    toggleExpandAll = () => {
        const {
            expandData,
            isExpandAll
        } = this.state;
        const {
            onExpandChange
        } = this.props;

        const newExpandData = {};

        // Expand/Collapse all rows
        Object.keys(expandData).forEach((rowId) => newExpandData[rowId] = !isExpandAll);

        // Callback function
        onExpandChange(newExpandData);

        this.setState({
            expandData: newExpandData,
            isExpandAll: !isExpandAll
        });
    }

    render() {
        const {
            children,
            isExpandAll: isExpandAllProp,
            onExpandChange,
            expandData: expandDataProp,
            ...rest
        } = this.props;
        const {
            expandData,
            isExpandAll
        } = this.state;

        return (
            <div {...rest}>
                {React.Children.toArray(children).map(child => {
                    const isTreeHead = child.type && child.type.displayName === 'TreeView.Head';
                    const isTree = child.type && child.type.displayName === 'TreeView.Tree';

                    if (isTreeHead) {
                        // Pass expand all callbacks to TreeHead
                        return React.cloneElement(child, {
                            onExpandAll: this.toggleExpandAll,
                            isExpanded: isExpandAll
                        });
                    }

                    if (isTree) {
                        // Pass expand callbacks to Tree
                        return React.cloneElement(child, {
                            expandData,
                            onExpandClick: this.toggleExpand
                        });
                    }

                    return child;
                })}
            </div>
        );
    }
}

TreeView.displayName = 'TreeView';

TreeView.propTypes = {
    children: PropTypes.node,
    expandData: PropTypes.object,
    isExpandAll: PropTypes.bool,
    onExpandChange: PropTypes.func
};

TreeView.defaultProps = {
    onExpandChange: () => {}
};

TreeView.propDescriptions = {
    children: 'Node(s) to render within the component. Expecting a `TreeHead` and a `Tree` component as children.',
    expandData: 'Object with rowId keys and boolean values representing whether that row is expanded. This variable is handled internally, but can be overridden by the consumer through this prop.',
    isExpandAll: 'Set to *true* for an expanded tree. This variable is handled internally, but can be overridden by the consumer through this prop',
    onExpandChange: 'Callback that is called whenever the internal expand/collapse state changes. The argument is an an object with rowId keys and boolean values representing whether that row is expanded.'
};

TreeView.Tree = Tree;
TreeView.Branch = TreeBranch;
TreeView.Col = TreeCol;
TreeView.Head = TreeHead;
TreeView.Item = TreeItem;
TreeView.Row = TreeRow;

export default TreeView;
