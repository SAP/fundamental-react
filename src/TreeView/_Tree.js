import BaseTree from './_BaseTree';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

class Tree extends Component {
    render() {
        return <BaseTree {...this.props} level={0} />;
    }
}

Tree.displayName = 'TreeView.Tree';

Tree.propTypes = {
    children: PropTypes.node,
    expandData: PropTypes.object,
    isExpanded: PropTypes.bool,
    onExpandClick: PropTypes.func
};

Tree.defaultProps = {
    expandData: {}
};

Tree.propDescriptions = {
    children: 'Node(s) to render within the component. Expecting `TreeItem` components as children.',
    expandData: '_INTERNAL USE ONLY._',
    isExpanded: '_INTERNAL USE ONLY._',
    level: '_INTERNAL USE ONLY._',
    onExpandClick: '_INTERNAL USE ONLY._'
};

export default Tree;
