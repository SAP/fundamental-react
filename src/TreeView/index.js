export { default as TreeView } from './TreeView.js';
