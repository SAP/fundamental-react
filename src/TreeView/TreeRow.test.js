import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import TreeView from './TreeView';

// mock shortid for snapshot testing
jest.mock('shortid', () => {
    let id = 1;

    return {
        generate: () => id++
    };
});

describe('<TreeView.Row />', () => {
    const multiLevelTreeView = (
        <TreeView>
            <TreeView.Head>
                <TreeView.Col>Column Header 1</TreeView.Col>
                <TreeView.Col>Column Header 2</TreeView.Col>
                <TreeView.Col>Column Header 3</TreeView.Col>
                <TreeView.Col>Column Header 4</TreeView.Col>
            </TreeView.Head>
            <TreeView.Tree>
                <TreeView.Item>
                    <TreeView.Row>
                        <TreeView.Col>First Level</TreeView.Col>
                        <TreeView.Col>Data Col 2</TreeView.Col>
                        <TreeView.Col>Data Col 3</TreeView.Col>
                        <TreeView.Col>Data Col 4</TreeView.Col>
                    </TreeView.Row>
                    <TreeView.Branch>
                        <TreeView.Item>
                            <TreeView.Row>
                                <TreeView.Col>Child 1</TreeView.Col>
                                <TreeView.Col>Data Col 2</TreeView.Col>
                                <TreeView.Col>Data Col 3</TreeView.Col>
                                <TreeView.Col>Data Col 4</TreeView.Col>
                            </TreeView.Row>
                            <TreeView.Branch>
                                <TreeView.Item>
                                    <TreeView.Row>
                                        <TreeView.Col>Grandchild 1</TreeView.Col>
                                        <TreeView.Col>Data Col 2</TreeView.Col>
                                        <TreeView.Col>Data Col 3</TreeView.Col>
                                        <TreeView.Col>Data Col 4</TreeView.Col>
                                    </TreeView.Row>
                                    <TreeView.Branch>
                                        <TreeView.Item>
                                            <TreeView.Row>
                                                <TreeView.Col>GreatGrandchild 1</TreeView.Col>
                                                <TreeView.Col>Data Col 2</TreeView.Col>
                                                <TreeView.Col>Data Col 3</TreeView.Col>
                                                <TreeView.Col>Data Col 4</TreeView.Col>
                                            </TreeView.Row>
                                        </TreeView.Item>
                                    </TreeView.Branch>
                                </TreeView.Item>
                            </TreeView.Branch>
                        </TreeView.Item>
                        <TreeView.Item>
                            <TreeView.Row>
                                <TreeView.Col>Child 2</TreeView.Col>
                                <TreeView.Col>Data Col 2</TreeView.Col>
                                <TreeView.Col>Data Col 3</TreeView.Col>
                                <TreeView.Col>Data Col 4</TreeView.Col>
                            </TreeView.Row>
                        </TreeView.Item>
                    </TreeView.Branch>
                </TreeView.Item>
                <TreeView.Item>
                    <TreeView.Row>
                        <TreeView.Col>Row 2</TreeView.Col>
                        <TreeView.Col>Data Col 2</TreeView.Col>
                        <TreeView.Col>Data Col 3</TreeView.Col>
                        <TreeView.Col>Data Col 4</TreeView.Col>
                    </TreeView.Row>
                    <TreeView.Branch>
                        <TreeView.Item>
                            <TreeView.Row>
                                <TreeView.Col>Child 1</TreeView.Col>
                                <TreeView.Col>Data Col 2</TreeView.Col>
                                <TreeView.Col>Data Col 3</TreeView.Col>
                                <TreeView.Col>Data Col 4</TreeView.Col>
                            </TreeView.Row>
                        </TreeView.Item>
                        <TreeView.Item>
                            <TreeView.Row>
                                <TreeView.Col>Child 2</TreeView.Col>
                                <TreeView.Col>Data Col 2</TreeView.Col>
                                <TreeView.Col>Data Col 3</TreeView.Col>
                                <TreeView.Col>Data Col 4</TreeView.Col>
                            </TreeView.Row>
                        </TreeView.Item>
                    </TreeView.Branch>
                </TreeView.Item>
                <TreeView.Item>
                    <TreeView.Row>
                        <TreeView.Col>Row 3</TreeView.Col>
                        <TreeView.Col>Data Col 2</TreeView.Col>
                        <TreeView.Col>Data Col 3</TreeView.Col>
                        <TreeView.Col>Data Col 4</TreeView.Col>
                    </TreeView.Row>
                </TreeView.Item>
                <TreeView.Item>
                    <TreeView.Row>
                        <TreeView.Col><a href='http://me.com' /></TreeView.Col>
                        <TreeView.Col>Data Col 2</TreeView.Col>
                        <TreeView.Col>Data Col 3</TreeView.Col>
                        <TreeView.Col>Data Col 4</TreeView.Col>
                    </TreeView.Row>
                    <TreeView.Branch>
                        <TreeView.Item>
                            <TreeView.Row>
                                <TreeView.Col>Child 1</TreeView.Col>
                                <TreeView.Col>Data Col 2</TreeView.Col>
                                <TreeView.Col>Data Col 3</TreeView.Col>
                                <TreeView.Col>Data Col 4</TreeView.Col>
                            </TreeView.Row>
                        </TreeView.Item>
                    </TreeView.Branch>
                </TreeView.Item>
            </TreeView.Tree>
        </TreeView>
    );

    const richTreeView = (
        <TreeView>
            <TreeView.Head>
                <TreeView.Col>Column Header 1</TreeView.Col>
                <TreeView.Col>Column Header 2</TreeView.Col>
                <TreeView.Col>Column Header 3</TreeView.Col>
                <TreeView.Col>Column Header 4</TreeView.Col>
            </TreeView.Head>
            <TreeView.Tree>
                <TreeView.Item>
                    <TreeView.Row>
                        <TreeView.Col>First Level</TreeView.Col>
                    </TreeView.Row>
                    <TreeView.Branch>
                        <TreeView.Item>
                            <TreeView.Row>
                                <TreeView.Col><a href='http://me.com'>First Level</a></TreeView.Col>
                                <TreeView.Col />
                                <TreeView.Col />
                                <TreeView.Col />
                            </TreeView.Row>
                            <TreeView.Branch>
                                <TreeView.Item>
                                    <TreeView.Row>
                                        <TreeView.Col><a href='http://amazon.com'>Amazon</a></TreeView.Col>
                                        <TreeView.Col><a href='http://google.com'>Google</a></TreeView.Col>
                                        <TreeView.Col><a href='http://bing.com'>Bing</a></TreeView.Col>
                                        <TreeView.Col><a href='http://yahoo.com'>Yahoo</a></TreeView.Col>
                                        <TreeView.Col><a href='http://sap.com'>SAP</a></TreeView.Col>
                                    </TreeView.Row>
                                </TreeView.Item>
                                <TreeView.Item>
                                    <TreeView.Row>
                                        <TreeView.Col><a href='http://amazon.com'>Amazon</a></TreeView.Col>
                                    </TreeView.Row>
                                </TreeView.Item>
                            </TreeView.Branch>
                        </TreeView.Item>
                        <TreeView.Item>
                            <TreeView.Row>
                                <TreeView.Col>Second Level</TreeView.Col>
                                <TreeView.Col />
                                <TreeView.Col />
                                <TreeView.Col />
                            </TreeView.Row>
                            <TreeView.Branch>
                                <TreeView.Item>
                                    <TreeView.Row>
                                        <TreeView.Col><a href='http://amazon.com'>Amazon</a></TreeView.Col>
                                    </TreeView.Row>
                                </TreeView.Item>
                            </TreeView.Branch>
                        </TreeView.Item>
                    </TreeView.Branch>
                </TreeView.Item>
            </TreeView.Tree>
        </TreeView>
    );

    const falseyCondition = false;
    const conditionalTreeView = (
        <TreeView>
            <TreeView.Head>
                <TreeView.Col>Column Header 1</TreeView.Col>
                {falseyCondition && <TreeView.Col>Column Header 2</TreeView.Col>}
                <TreeView.Col>Column Header 3</TreeView.Col>
                <TreeView.Col>Column Header 4</TreeView.Col>
            </TreeView.Head>
            <TreeView.Tree>
                <TreeView.Item>
                    <TreeView.Row>
                        <TreeView.Col>First Level</TreeView.Col>
                    </TreeView.Row>
                    <TreeView.Branch>
                        <TreeView.Item>
                            <TreeView.Row>
                                {falseyCondition && (
                                    <TreeView.Col>
                                        <a href='http://me.com'>First Level</a>
                                    </TreeView.Col>
                                )}
                                <TreeView.Col>Second level</TreeView.Col>
                                <TreeView.Col />
                                <TreeView.Col />
                            </TreeView.Row>
                            {falseyCondition && <TreeView.Branch />}
                        </TreeView.Item>
                        {falseyCondition && <TreeView.Item />}
                    </TreeView.Branch>
                </TreeView.Item>
                {falseyCondition && <TreeView.Item />}
            </TreeView.Tree>
        </TreeView>
    );

    test('create tree component', () => {
        // multi-level tree
        let component = renderer.create(multiLevelTreeView);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // rich tree
        component = renderer.create(richTreeView);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // conditional tree
        component = renderer.create(conditionalTreeView);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('open all tree from header', () => {
        const wrapper = mount(richTreeView);

        wrapper
            .find('button.fd-tree__control')
            .at(0)
            .simulate('click');

        // Check that all opened
        wrapper.find('ul.fd-tree__group').forEach((node) => {
            expect(node.hasClass('is-hidden')).toBeFalsy();
            expect(node.prop('aria-hidden')).toBeFalsy();
        });
    });

    test('closed all tree from header when set to open', () => {
        const wrapper = mount(richTreeView);

        wrapper
            .find('button.fd-tree__control')
            .at(0)
            .simulate('click');

        // Check that all opened
        wrapper.find('ul.fd-tree__group').forEach((node) => {
            expect(node.hasClass('is-hidden')).toBeFalsy();
            expect(node.prop('aria-hidden')).toBeFalsy();
        });

        wrapper
            .find('button.fd-tree__control')
            .at(0)
            .simulate('click');

        // Check that all closed
        wrapper.find('ul.fd-tree__group').forEach((node) => {
            expect(node.hasClass('is-hidden')).toBeTruthy();
            expect(node.prop('aria-hidden')).toBeTruthy();
        });
    });

    test('expand tree from row', () => {
        const wrapper = mount(multiLevelTreeView);
        const parent = wrapper.find('li.fd-tree__item').at(0);
        const button = parent.find('button.fd-tree__control').at(0);

        button.simulate('click');

        // Check that child list expanded
        expect(wrapper.find('ul.fd-tree__group--sublevel-1').at(0).hasClass('is-hidden')).toBeFalsy();
        expect(wrapper.find('ul.fd-tree__group--sublevel-1').at(0).prop('aria-hidden')).toBeFalsy();
    });

    test('collapse tree from row', () => {
        const wrapper = mount(multiLevelTreeView);
        const parent = wrapper.find('li.fd-tree__item').at(0);
        const button = parent.find('button.fd-tree__control').at(0);

        button.simulate('click');

        // Check that child list expanded
        expect(wrapper.find('ul.fd-tree__group--sublevel-1').at(0).hasClass('is-hidden')).toBeFalsy();
        expect(wrapper.find('ul.fd-tree__group--sublevel-1').at(0).prop('aria-hidden')).toBeFalsy();

        button.simulate('click');

        // Check that child list collapsed
        expect(wrapper.find('ul.fd-tree__group--sublevel-1').at(0).hasClass('is-hidden')).toBeTruthy();
        expect(wrapper.find('ul.fd-tree__group--sublevel-1').at(0).prop('aria-hidden')).toBeTruthy();
    });

    test('expand change callback is called with correct arguments on row expand', () => {
        const onExpandChange = jest.fn();
        const callbackTree = (
            <TreeView onExpandChange={onExpandChange}>
                <TreeView.Tree>
                    <TreeView.Item rowId={'testRowId'}>
                        <TreeView.Row>
                            <TreeView.Col>Row 1</TreeView.Col>
                        </TreeView.Row>
                        <TreeView.Branch>
                            <TreeView.Item>
                                <TreeView.Row>
                                    <TreeView.Col>Row 2</TreeView.Col>
                                </TreeView.Row>
                            </TreeView.Item>
                        </TreeView.Branch>
                    </TreeView.Item>
                </TreeView.Tree>
            </TreeView>
        );
        const element = mount(callbackTree);
        const button = element.find('button').at(0);

        // Simulate clicking the expand row button
        button.simulate('click');

        expect(onExpandChange).toHaveBeenLastCalledWith(expect.objectContaining({ testRowId: true }));
    });

    test('expand change callback is called with correct arguments on header expand', () => {
        const onExpandChange = jest.fn();
        const callbackTree = (
            <TreeView onExpandChange={onExpandChange}>
                <TreeView.Head>
                    <TreeView.Col>Header 1</TreeView.Col>
                </TreeView.Head>
                <TreeView.Tree>
                    <TreeView.Item>
                        <TreeView.Row>
                            <TreeView.Col>Row 1</TreeView.Col>
                        </TreeView.Row>
                        <TreeView.Branch>
                            <TreeView.Item>
                                <TreeView.Row>
                                    <TreeView.Col>Row 2</TreeView.Col>
                                </TreeView.Row>
                            </TreeView.Item>
                        </TreeView.Branch>
                    </TreeView.Item>
                </TreeView.Tree>
            </TreeView>
        );
        const element = mount(callbackTree);
        const button = element.find('button').at(0);

        // Simulate clicking the expand all button
        button.simulate('click');

        const lastCallArg = onExpandChange.mock.calls[onExpandChange.mock.calls.length - 1][0];

        // Expect all expand values to be true
        Object.keys(lastCallArg).forEach((id) => {
            expect(lastCallArg[id]).toBeTruthy();
        });
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the TreeView component', () => {
            const element = mount(
                <TreeView
                    data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the TreeItem component', () => {
            const element = mount(
                <TreeView.Item
                    data-sample='Sample' />);

            expect(
                element.find('.fd-tree__item').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the TreeRow component', () => {
            const element = mount(
                <TreeView.Row
                    data-sample='Sample' />);

            expect(
                element.find('.fd-tree__row').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });
});
