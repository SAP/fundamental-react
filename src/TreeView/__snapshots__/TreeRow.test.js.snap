// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TreeView.Row /> create tree component 1`] = `
<div>
  <div
    className="fd-tree fd-tree--header"
  >
    <div
      className="fd-tree__row fd-tree__row--header"
    >
      <div
        className="fd-tree__col fd-tree__col--control"
      >
        <div>
          <button
            aria-label="expand all"
            aria-pressed={false}
            className="fd-tree__control"
            onClick={[Function]}
          />
          Column Header 1
        </div>
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 2
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 3
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 4
      </div>
    </div>
  </div>
  <ul
    aria-hidden={false}
    className="fd-tree"
    role="tree"
  >
    <li
      aria-expanded={false}
      className="fd-tree__item"
      id="fd-1"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <div>
            <button
              aria-controls="fd-1"
              aria-label="expand"
              aria-pressed={false}
              className="fd-tree__control"
              onClick={[Function]}
            />
            First Level
          </div>
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 2
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 3
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 4
        </div>
      </div>
      <ul
        aria-hidden={true}
        className="fd-tree__group fd-tree__group--sublevel-1 is-hidden"
        role="group"
      >
        <li
          aria-expanded={false}
          className="fd-tree__item"
          id="fd-2"
          role="treeitem"
        >
          <div
            className="fd-tree__row"
          >
            <div
              className="fd-tree__col fd-tree__col--control"
            >
              <div>
                <button
                  aria-controls="fd-2"
                  aria-label="expand"
                  aria-pressed={false}
                  className="fd-tree__control"
                  onClick={[Function]}
                />
                Child 1
              </div>
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 2
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 3
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 4
            </div>
          </div>
          <ul
            aria-hidden={true}
            className="fd-tree__group fd-tree__group--sublevel-2 is-hidden"
            role="group"
          >
            <li
              aria-expanded={false}
              className="fd-tree__item"
              id="fd-3"
              role="treeitem"
            >
              <div
                className="fd-tree__row"
              >
                <div
                  className="fd-tree__col fd-tree__col--control"
                >
                  <div>
                    <button
                      aria-controls="fd-3"
                      aria-label="expand"
                      aria-pressed={false}
                      className="fd-tree__control"
                      onClick={[Function]}
                    />
                    Grandchild 1
                  </div>
                </div>
                <div
                  className="fd-tree__col"
                >
                  Data Col 2
                </div>
                <div
                  className="fd-tree__col"
                >
                  Data Col 3
                </div>
                <div
                  className="fd-tree__col"
                >
                  Data Col 4
                </div>
              </div>
              <ul
                aria-hidden={true}
                className="fd-tree__group fd-tree__group--sublevel-3 is-hidden"
                role="group"
              >
                <li
                  aria-expanded={false}
                  className="fd-tree__item"
                  id="fd-4"
                  role="treeitem"
                >
                  <div
                    className="fd-tree__row"
                  >
                    <div
                      className="fd-tree__col fd-tree__col--control"
                    >
                      GreatGrandchild 1
                    </div>
                    <div
                      className="fd-tree__col"
                    >
                      Data Col 2
                    </div>
                    <div
                      className="fd-tree__col"
                    >
                      Data Col 3
                    </div>
                    <div
                      className="fd-tree__col"
                    >
                      Data Col 4
                    </div>
                  </div>
                </li>
              </ul>
            </li>
          </ul>
        </li>
        <li
          aria-expanded={false}
          className="fd-tree__item"
          id="fd-5"
          role="treeitem"
        >
          <div
            className="fd-tree__row"
          >
            <div
              className="fd-tree__col fd-tree__col--control"
            >
              Child 2
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 2
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 3
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 4
            </div>
          </div>
        </li>
      </ul>
    </li>
    <li
      aria-expanded={false}
      className="fd-tree__item"
      id="fd-6"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <div>
            <button
              aria-controls="fd-6"
              aria-label="expand"
              aria-pressed={false}
              className="fd-tree__control"
              onClick={[Function]}
            />
            Row 2
          </div>
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 2
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 3
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 4
        </div>
      </div>
      <ul
        aria-hidden={true}
        className="fd-tree__group fd-tree__group--sublevel-1 is-hidden"
        role="group"
      >
        <li
          aria-expanded={false}
          className="fd-tree__item"
          id="fd-7"
          role="treeitem"
        >
          <div
            className="fd-tree__row"
          >
            <div
              className="fd-tree__col fd-tree__col--control"
            >
              Child 1
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 2
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 3
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 4
            </div>
          </div>
        </li>
        <li
          aria-expanded={false}
          className="fd-tree__item"
          id="fd-8"
          role="treeitem"
        >
          <div
            className="fd-tree__row"
          >
            <div
              className="fd-tree__col fd-tree__col--control"
            >
              Child 2
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 2
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 3
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 4
            </div>
          </div>
        </li>
      </ul>
    </li>
    <li
      aria-expanded={false}
      className="fd-tree__item"
      id="fd-9"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          Row 3
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 2
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 3
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 4
        </div>
      </div>
    </li>
    <li
      aria-expanded={false}
      className="fd-tree__item"
      id="fd-10"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <div>
            <button
              aria-controls="fd-10"
              aria-label="expand"
              aria-pressed={false}
              className="fd-tree__control"
              onClick={[Function]}
            />
            <a
              href="http://me.com"
            />
          </div>
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 2
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 3
        </div>
        <div
          className="fd-tree__col"
        >
          Data Col 4
        </div>
      </div>
      <ul
        aria-hidden={true}
        className="fd-tree__group fd-tree__group--sublevel-1 is-hidden"
        role="group"
      >
        <li
          aria-expanded={false}
          className="fd-tree__item"
          id="fd-11"
          role="treeitem"
        >
          <div
            className="fd-tree__row"
          >
            <div
              className="fd-tree__col fd-tree__col--control"
            >
              Child 1
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 2
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 3
            </div>
            <div
              className="fd-tree__col"
            >
              Data Col 4
            </div>
          </div>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`<TreeView.Row /> create tree component 2`] = `
<div>
  <div
    className="fd-tree fd-tree--header"
  >
    <div
      className="fd-tree__row fd-tree__row--header"
    >
      <div
        className="fd-tree__col fd-tree__col--control"
      >
        <div>
          <button
            aria-label="expand all"
            aria-pressed={false}
            className="fd-tree__control"
            onClick={[Function]}
          />
          Column Header 1
        </div>
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 2
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 3
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 4
      </div>
    </div>
  </div>
  <ul
    aria-hidden={false}
    className="fd-tree"
    role="tree"
  >
    <li
      aria-expanded={false}
      className="fd-tree__item"
      id="fd-12"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <div>
            <button
              aria-controls="fd-12"
              aria-label="expand"
              aria-pressed={false}
              className="fd-tree__control"
              onClick={[Function]}
            />
            First Level
          </div>
        </div>
      </div>
      <ul
        aria-hidden={true}
        className="fd-tree__group fd-tree__group--sublevel-1 is-hidden"
        role="group"
      >
        <li
          aria-expanded={false}
          className="fd-tree__item"
          id="fd-13"
          role="treeitem"
        >
          <div
            className="fd-tree__row"
          >
            <div
              className="fd-tree__col fd-tree__col--control"
            >
              <div>
                <button
                  aria-controls="fd-13"
                  aria-label="expand"
                  aria-pressed={false}
                  className="fd-tree__control"
                  onClick={[Function]}
                />
                <a
                  href="http://me.com"
                >
                  First Level
                </a>
              </div>
            </div>
            <div
              className="fd-tree__col"
            />
            <div
              className="fd-tree__col"
            />
            <div
              className="fd-tree__col"
            />
          </div>
          <ul
            aria-hidden={true}
            className="fd-tree__group fd-tree__group--sublevel-2 is-hidden"
            role="group"
          >
            <li
              aria-expanded={false}
              className="fd-tree__item"
              id="fd-14"
              role="treeitem"
            >
              <div
                className="fd-tree__row"
              >
                <div
                  className="fd-tree__col fd-tree__col--control"
                >
                  <a
                    href="http://amazon.com"
                  >
                    Amazon
                  </a>
                </div>
                <div
                  className="fd-tree__col"
                >
                  <a
                    href="http://google.com"
                  >
                    Google
                  </a>
                </div>
                <div
                  className="fd-tree__col"
                >
                  <a
                    href="http://bing.com"
                  >
                    Bing
                  </a>
                </div>
                <div
                  className="fd-tree__col"
                >
                  <a
                    href="http://yahoo.com"
                  >
                    Yahoo
                  </a>
                </div>
                <div
                  className="fd-tree__col"
                >
                  <a
                    href="http://sap.com"
                  >
                    SAP
                  </a>
                </div>
              </div>
            </li>
            <li
              aria-expanded={false}
              className="fd-tree__item"
              id="fd-15"
              role="treeitem"
            >
              <div
                className="fd-tree__row"
              >
                <div
                  className="fd-tree__col fd-tree__col--control"
                >
                  <a
                    href="http://amazon.com"
                  >
                    Amazon
                  </a>
                </div>
              </div>
            </li>
          </ul>
        </li>
        <li
          aria-expanded={false}
          className="fd-tree__item"
          id="fd-16"
          role="treeitem"
        >
          <div
            className="fd-tree__row"
          >
            <div
              className="fd-tree__col fd-tree__col--control"
            >
              <div>
                <button
                  aria-controls="fd-16"
                  aria-label="expand"
                  aria-pressed={false}
                  className="fd-tree__control"
                  onClick={[Function]}
                />
                Second Level
              </div>
            </div>
            <div
              className="fd-tree__col"
            />
            <div
              className="fd-tree__col"
            />
            <div
              className="fd-tree__col"
            />
          </div>
          <ul
            aria-hidden={true}
            className="fd-tree__group fd-tree__group--sublevel-2 is-hidden"
            role="group"
          >
            <li
              aria-expanded={false}
              className="fd-tree__item"
              id="fd-17"
              role="treeitem"
            >
              <div
                className="fd-tree__row"
              >
                <div
                  className="fd-tree__col fd-tree__col--control"
                >
                  <a
                    href="http://amazon.com"
                  >
                    Amazon
                  </a>
                </div>
              </div>
            </li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`<TreeView.Row /> create tree component 3`] = `
<div>
  <div
    className="fd-tree fd-tree--header"
  >
    <div
      className="fd-tree__row fd-tree__row--header"
    >
      <div
        className="fd-tree__col fd-tree__col--control"
      >
        <div>
          <button
            aria-label="expand all"
            aria-pressed={false}
            className="fd-tree__control"
            onClick={[Function]}
          />
          Column Header 1
        </div>
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 3
      </div>
      <div
        className="fd-tree__col"
      >
        Column Header 4
      </div>
    </div>
  </div>
  <ul
    aria-hidden={false}
    className="fd-tree"
    role="tree"
  >
    <li
      aria-expanded={false}
      className="fd-tree__item"
      id="fd-18"
      role="treeitem"
    >
      <div
        className="fd-tree__row"
      >
        <div
          className="fd-tree__col fd-tree__col--control"
        >
          <div>
            <button
              aria-controls="fd-18"
              aria-label="expand"
              aria-pressed={false}
              className="fd-tree__control"
              onClick={[Function]}
            />
            First Level
          </div>
        </div>
      </div>
      <ul
        aria-hidden={true}
        className="fd-tree__group fd-tree__group--sublevel-1 is-hidden"
        role="group"
      >
        <li
          aria-expanded={false}
          className="fd-tree__item"
          id="fd-19"
          role="treeitem"
        >
          <div
            className="fd-tree__row"
          >
            <div
              className="fd-tree__col fd-tree__col--control"
            >
              Second level
            </div>
            <div
              className="fd-tree__col"
            />
            <div
              className="fd-tree__col"
            />
          </div>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
